### PR TITLE
Examples for Config builder region function

### PR DESCRIPTION
**Pull request by Jacco**
## Motivation and Context
#670 

## Description
Added documentation to the region function in `RegionDecorator.kt`.
Also had to swap a few parameters around.

## Testing

`./gradlew :aws:sdk:assemble`
`cd aws/sdk/build/aws-sdk/sdk/dynamodb`
`cargo doc --open`

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Jacco** [comment]
@rcoh shall I add the method of testing the docs to project readme?

**Comment by rcoh** [comment]
Sounds good! May even be worth splitting of a developer guide of some sort

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -1,46 +1,5 @@
...
-# message = "Fix typos in module documentation for generated crates"
-# references = ["smithy-rs#920"]
-# meta = { "breaking" = false, "tada" = false, "bug" = false }
-# author = "rcoh"
-#
-# [[smithy-rs]]
-# message = "Fix typos in module documentation for generated crates"
-# references = ["smithy-rs#920"]
-# meta = { "breaking" = false, "tada" = false, "bug" = false }
-# author = "rcoh"
```
Comment (rcoh) [inline CHANGELOG.next.toml:12]: something went wrong merging the changelog. Can you leave the example changelog entries there as well?

**Review by rcoh — COMMENTED** [review]
changes look good, although changing the constructor may break an existing test.

File: CHANGELOG.next.toml (hunk 1 above)
Comment (Jacco) [reply CHANGELOG.next.toml]: I have put the examples back

**Comment by Jacco** [comment]
> changes look good, although changing the constructor may break an existing test: https://github.com/awslabs/smithy-rs/blob/72a5c5b3e8bd6f722aade6aafb180b51d5f36ac4/aws/sdk-codegen/src/test/kotlin/software/amazon/smithy/rustsdk/RegionProviderConfigTest.kt

I have a hard time figuring if all the test run good. When I run `./gradlew :aws:sdk:test` there is always something red.

Found a use error in the docs. Did not find the constructor test failure yet.

**Comment by Jacco** [comment]
Corrently I get this one:

```
 Documenting aws-types v0.3.0 (/Users/jaccokulman/Desktop/smithy-rs/aws/sdk/build/aws-sdk/sdk/aws-types)
error: lint `missing_crate_level_docs` has been renamed to `rustdoc::missing_crate_level_docs`
  --> sdk/aws-types/src/lib.rs:10:5
   |
10 |     missing_crate_level_docs,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `rustdoc::missing_crate_level_docs`
   |
   = note: `-D renamed-and-removed-lints` implied by `-D warnings`

error: Compilation failed, aborting rustdoc

error: could not document `aws-types`

Caused by:
  process didn't exit successfully: `rustdoc --edition=2018 --crate-type lib --crate-name aws_types sdk/aws-types/src/lib.rs -o /Users/jaccokulman/Desktop/smithy-rs/target/doc --error-format=json --json=diagnostic-rendered-ansi --document-private-items -L dependency=/Users/jaccokulman/Desktop/smithy-rs/target/debug/deps --extern aws_smithy_async=/Users/jaccokulman/Desktop/smithy-rs/target/debug/deps/libaws_smithy_async-424b745062c3df04.rmeta --extern aws_smithy_types=/Users/jaccokulman/Desktop/smithy-rs/target/debug/deps/libaws_smithy_types-30f3078dec8f7bb6.rmeta --extern tracing=/Users/jaccokulman/Desktop/smithy-rs/target/debug/deps/libtracing-55c587832def698a.rmeta --extern zeroize=/Users/jaccokulman/Desktop/smithy-rs/target/debug/deps/libzeroize-ec7e5a396ea5bd7d.rmeta -D warnings --crate-version 0.3.0` (exit status: 1)
warning: build failed, waiting for other jobs to finish...
error: build failed

> Task :aws:sdk:cargoDocs FAILED
```

The `missing_crate_level_docs` scolls by a lot when executing the tests but all the other times it is a warning.

And it reports the error occurring in `aws-types` of which I did not touch the docs.

**Comment by rcoh** [comment]
what rustc version are you running? I think that's causing the rustdoc changes.

For fixing the unit test, `./gradlew :aws:sdk-codegen:test` to run the unit tests in the sdk-codegen directory

**Comment by rcoh** [comment]
For fixing the test, a small refactoring is required since we need a test codegen context:
1. Add this to `TestUtil.kt` in `sdk-codegen`:
```kotlin
fun awsTestCodegenContext(model: Model? = null) =
    testCodegenContext(model ?: "namespace test".asSmithyModel()).copy(runtimeConfig = AwsTestRuntimeConfig)
```

2. In the test, instead of passing AwsTestRuntimeConfig, pass `awsTestCodegenContext` instead

**Comment by rcoh** [comment]
For fixing the test, a small refactoring is required since we need a test codegen context:
1. Add this to `TestUtil.kt` in `sdk-codegen`:
```kotlin
fun awsTestCodegenContext(model: Model? = null) =
    testCodegenContext(model ?: "namespace test".asSmithyModel()).copy(runtimeConfig = AwsTestRuntimeConfig)
```

2. In the test, instead of passing AwsTestRuntimeConfig, pass `awsTestCodegenContext` instead

**Comment by rcoh** [comment]
For fixing the test, a small refactoring is required since we need a test codegen context:
1. Add this to `TestUtil.kt` in `sdk-codegen`:
```kotlin
fun awsTestCodegenContext(model: Model? = null) =
    testCodegenContext(model ?: "namespace test".asSmithyModel()).copy(runtimeConfig = AwsTestRuntimeConfig)
```

2. In the test, instead of passing AwsTestRuntimeConfig, pass `awsTestCodegenContext` instead

**Comment by Jacco** [comment]
Thx for the tips! I did applied your changes but still get an error executing the tests:

```
software.amazon.smithy.rustsdk.RegionProviderConfigTest > generates a valid config() FAILED
    software.amazon.smithy.rust.codegen.util.CommandFailed at RegionProviderConfigTest.kt:14
```

When running with info:

```
software.amazon.smithy.rustsdk.RegionProviderConfigTest > generates a valid config() FAILED
    CommandFailed(output=Unexpected exception thrown when executing subprocess:
    CommandFailed(output=Command Failed
    error: failed to load manifest for workspace member `/private/var/folders/k0/bphpmb9j5q75gnpdrfgtpd8m0000gp/T/smithy-test12470980131730359897/smithy-test8835940347322312888`

    Caused by:
      failed to parse manifest at `/private/var/folders/k0/bphpmb9j5q75gnpdrfgtpd8m0000gp/T/smithy-test12470980131730359897/smithy-test8835940347322312888/Cargo.toml`

    Caused by:
      no targets specified in the manifest
      either src/lib.rs, src/main.rs, a [lib] section, or [[bin]] section must be present

    ))
        at software.amazon.smithy.rust.codegen.util.ExecKt.runCommand(Exec.kt:41)
        at software.amazon.smithy.rust.codegen.util.ExecKt.runCommand$default(Exec.kt:15)
        at software.amazon.smithy.rust.codegen.testutil.RustKt.compileAndTest(Rust.kt:185)
        at software.amazon.smithy.rust.codegen.testutil.RustKt.compileAndTest$default(Rust.kt:167)
        at software.amazon.smithy.rust.codegen.testutil.TestConfigCustomizationKt.validateConfigCustomizations(TestConfigCustomization.kt:53)
        at software.amazon.smithy.rust.codegen.testutil.TestConfigCustomizationKt.validateConfigCustomizations$default(TestConfigCustomization.kt:49)
        at software.amazon.smithy.rustsdk.RegionProviderConfigTest.generates a valid config(RegionProviderConfigTest.kt:39)
```

Also flowwing tests seem to fail after this initial error.

My rustc version is

`rustc 1.56.1 (59eed8a2a 2021-11-01)`

File: CHANGELOG.next.toml [hunk 2]
```diff
@@ -12,35 +12,7 @@
...
-references = ["smithy-rs#998", "aws-sdk-rust#359"]
-author = "rcoh"
-
-[[aws-sdk-rust]]
-message = "aws_types::Config is now `Clone`"
+message = "Example for Config builder region function added"
+references = ["smithy-rs#670"]
 meta = { "breaking" = false, "tada" = false, "bug" = false }
-references = ["smithy-rs#1002", "aws-sdk-rust#352"]
-author = "rcoh"
```
Comment (rcoh) [inline CHANGELOG.next.toml:46]: if you look at this diff, you can see that your changelog update deleted other changelog entries

**Comment by rcoh** [comment]
can you push your changes? I can pull them down locally and push a commit to get the tests passing

File: CHANGELOG.next.toml (hunk 2 above)
Comment (Jacco) [reply CHANGELOG.next.toml:46]: Ah sorry, I messed up my branch before I thought these  came from the mess-up. I will try put them back

**Comment by Jacco** [comment]
@rcoh pushed the changes

**Comment by rcoh** [comment]
I think the test failures are something to do with your local setup, everything is passing on CI :-)

**Review by rcoh — APPROVED** [review]
LGTM! thanks for taking this on!

**Comment by rcoh** [comment]
ah I found the issue, pushed a fix.

**Comment by Jacco** [comment]
Great :-)

**Pull request merged**